feat: function to collect bunch across MPI ranks into a single Python dictionary

woodtp opened the pull request:
This PR adds `collect_bunch` to `orbit.bunch_utils`, which accepts a `Bunch` and collects attributes across MPI ranks into a Python dictionary. The 6 coordinates (x, px, y, py, z, dE) are stored in the dictionary at key `"coords"` in an `np.ndarray` with shape `(N, 6)`. In the case of either attributes related to the synchronous particle or those contained in the bunch structure retrieved via `Bunch.bunchAttrDouble` or `Bunch.bunchAttrInt`, the bunch dictionary values are stored in their own dictionaries nested at keys `"sync_part"` and `"attributes"`, respectively, as either `np.float64` or `np.int32` where appropriate.

Since the output of `collect_bunch` is a dictionary of numpy objects, the user has the flexibility to pass directly to visualization libs and/or store the output in whichever format they wish. E.g.,

```python
from orbit.bunch_utils import collect_bunch

...

bunch_dict = collect_bunch(bunch)

if mpi_rank == 0:
    # plot x-x'
    sns.histplot(x=bunch_dict["coords"][:, 0], y=bunch_dict["coords"][:, 1])
    plt.savefig("nice_plot.png")
        
    # save flattened dict with np.savez or np.savez_compressed
    np.savez("filename.npz", coords=bunch_dict["coords"], **bunch_dict["sync_part"], **bunch_dict["attributes"])
    
    # Although if you save the coords separately as a `.npy` with `np.save`, you have the ability to 
    # load it later as a mmap, which is probably better for very large bunches.
    np.save("coords.npy", bunch_dict["coords"])
    
    # sometime later...
    coords = np.load("coords.npy", mmap_mode="r")
```

austin-hoover commented:
I just tested and it's working for me! One suggestion is to group the entire NX6 coordinate array into one entry called `coords` or `array` or something; is that possible?. Also would it make sense to have just one `sync_part` key and the value be another dictionary? @woodtp @azukov

woodtp commented:
> I just tested and it's working for me! One suggestion is to group the entire NX6 coordinate array into one entry called `coords` or `array` or something; is that possible?. Also would it make sense to have just one `sync_part` key and the value be another dictionary? @woodtp @azukov

I don't see why not, if that's preferable. That layout seems more consistent with [how a Bunch stores this information internally](https://github.com/PyORBIT-Collaboration/PyORBIT3/blob/6cc7421f641e7f9f68ba74059f332aeabdd8e0e9/src/orbit/Bunch.cc#L53C1-L56).

woodtp commented:
@azukov correctly pointed out to me that there was an issue with the implementation in the case where the global bunch size wasn't divisible by the total number of MPI nodes.

I have rewritten the utility entirely to resolve that issue while simultaneously avoiding loading the entire bunch into RAM, unless explicitly requested. It now works by creating a `np.memmap` backed by a file under `/tmp`, and each node writes to a specific slice of the memory-mapped array, in parallel.

I timed collecting a bunch with 100M particles using the new method, split across 7 nodes (to force an uneven distribution), and here are the results: 

```
# checking that the bunch gets distributed correctly and all `local_rows` add up to 100_000_000
[DEBUG] mpi_rank=0, fname='/tmp/collect_bunch_tmpfile.dat', start_row=0, local_rows=14_285_715
[DEBUG] mpi_rank=2, fname='/tmp/collect_bunch_tmpfile.dat', start_row=28_571_430, local_rows=14_285_714
[DEBUG] mpi_rank=3, fname='/tmp/collect_bunch_tmpfile.dat', start_row=42_857_144, local_rows=14_285_714
[DEBUG] mpi_rank=1, fname='/tmp/collect_bunch_tmpfile.dat', start_row=14_285_715, local_rows=14_285_715
[DEBUG] mpi_rank=4, fname='/tmp/collect_bunch_tmpfile.dat', start_row=57_142_858, local_rows=14_285_714
[DEBUG] mpi_rank=5, fname='/tmp/collect_bunch_tmpfile.dat', start_row=71_428_572, local_rows=14_285_714
[DEBUG] mpi_rank=6, fname='/tmp/collect_bunch_tmpfile.dat', start_row=85_714_286, local_rows=14_285_714

[DEBUG] (mpi_rank=1) COLLECTING BUNCH DONE in 34.941 sec
[DEBUG] (mpi_rank=3) COLLECTING BUNCH DONE in 34.941 sec
[DEBUG] (mpi_rank=6) COLLECTING BUNCH DONE in 34.942 sec
[DEBUG] (mpi_rank=2) COLLECTING BUNCH DONE in 34.944 sec
[DEBUG] (mpi_rank=5) COLLECTING BUNCH DONE in 34.944 sec
[DEBUG] (mpi_rank=4) COLLECTING BUNCH DONE in 34.944 sec
[DEBUG] (mpi_rank=0) COLLECTING BUNCH DONE in 34.990 sec

[DEBUG] coords size in bytes:  4800.0 MB 
[DEBUG] coords shape:  (100000000, 6)
```

woodtp commented on py/orbit/bunch_utils/collect_bunch.py:
```diff
@@ -0,0 +1,165 @@
+    #     file_desc, fname = tempfile.mkstemp(suffix=".dat", prefix="collect_bunch_", dir="/tmp")
+    #     os.close(file_desc)
+    #
+    # TODO: this doesn't seem to work. "SystemError: PY_SSIZE_T_CLEAN macro must be defined for '#' formats"
```
Note: Initially, I tried to use the built-in `tempfile` module to initialize the temporary file for the memmap. When attempting to broadcast the name of the file to the other nodes, it seems like I'm running into an instance of #65... So, I'm hard-coding the name of the file for now.

woodtp commented:
Ok, it still wasn't working properly after some particles in the bunch were lost and I was getting some fun artifacts when I tried plotting emittance spectra. I was able to fix that by writing bunch coords to separate mem-mapped arrays on each MPI node, then concatenating them on the primary rank at the end of the routine.

I also squashed the previous commits for readability.

azukov commented:
@woodtp  I think it's all good now. 
Maybe we want to merge _collect_bunch.py_ and _file_handler.py_ into _serialize.py_?
